src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +26,96 @@ public static class Renamer
+            Document document,
+            string newDocumentName,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentName == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentName));
+            }
+
+            var actions = new ArrayBuilder<RenameDocumentAction>();
+
+            if (!newDocumentName.Equals(document.Name))
+            {
+                var renameAction = await RenameSymbolDocumentAction.TryCreateAsync(document, newDocumentName, optionSet, cancellationToken).ConfigureAwait(false);
+
+                if (renameAction is object)
+                {
+                    actions.Add(renameAction);
+                }
+            }
+
+            return new RenameDocumentActionSet(
+                actions.ToImmutableAndFree(),
+                document.Project.Id,
+                document.Id,
+                newDocumentName,
+                document.Folders);
+        }
+
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithFolders(IEnumerable{string})" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="SyncNamespaceDocumentAction"/> that will sync the namespace(s) of the document to match the document folders
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentFoldersAsync(
+            Document document,
+            IReadOnlyList<string> newFolders,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newFolders == null)
+            {
+                throw new ArgumentNullException(nameof(newFolders));
+            }
+
+            var actions = new ArrayBuilder<RenameDocumentAction>();
+
+            if (!newFolders.SequenceEqual(document.Folders))
+            {
+                var action = await SyncNamespaceDocumentAction.TryCreateAsync(document, newFolders, optionSet, cancellationToken).ConfigureAwait(false);
+
+                if (action is object)
+                {
+                    actions.Add(action);
+                }
+            }
+
+            return new RenameDocumentActionSet(
```

Comment:
i don't like that you can get an empty set back.

Reply:
I got that the first time you said it ;) 

Closing to condense the feedback and have the discussion easier to follow. Using this thread for discussion https://github.com/dotnet/roslyn/pull/41267#discussion_r399520148